src/core/tsi/ssl_transport_security_utils.cc

```diff
@@ -375,4 +380,79 @@ absl::StatusOr<std::string> AkidFromCrl(X509_CRL* crl) {
   return ret;
 }
 
+bool IsPemCertificateChainNonEmptyAndValid(absl::string_view cert_chain_pem) {
+  if (cert_chain_pem.empty()) {
+    return false;
+  }
+  BIO* in = BIO_new_mem_buf(cert_chain_pem.data(), cert_chain_pem.size());
+  if (in == nullptr) {
+    return false;
+  }
+  uint8_t* cert_data = nullptr;
+  long cert_length = 0;
+  size_t number_of_valid_certs = 0;
+  while (PEM_bytes_read_bio(&cert_data, &cert_length, /*pnm=*/nullptr,
+                            PEM_STRING_X509, in, /*cb=*/nullptr,
+                            /*u=*/nullptr)) {
+    // Try to form the X.509 object from the DER-encoding.
+    const uint8_t* der = cert_data;
+    X509* x509 = d2i_X509(/*out=*/nullptr, &der, cert_length);
+    if (x509 == nullptr) {
+      BIO_free(in);
+      return false;
+    }
+    X509_free(x509);
+    OPENSSL_free(cert_data);
+    cert_data = nullptr;
+    cert_length = 0;
+    number_of_valid_certs++;
+  }
+  // We always have errors at this point because in the above loop we read until
+  // we reach the end of |cert_chain_pem|, which generates a "no start line"
+  // error. Therefore, this error is OK if we have successfully parsed some
+  // certificate data previously.
+  const int last_error = ERR_peek_last_error();
+  if (ERR_GET_LIB(last_error) != ERR_LIB_PEM ||
+      ERR_GET_REASON(last_error) != PEM_R_NO_START_LINE) {
+    BIO_free(in);
+    return false;
+  }
+  ERR_clear_error();
+  BIO_free(in);
+  return (number_of_valid_certs > 0);
+}
+
+bool IsRsaPrivateKey(absl::string_view private_key_pem) {
+  BIO* in = BIO_new_mem_buf(private_key_pem.data(), private_key_pem.size());
+  if (in == nullptr) {
+    return false;
+  }
+  RSA* rsa = PEM_read_bio_RSAPrivateKey(in, /*x=*/nullptr,
+                                        /*cb=*/nullptr, /*u=*/nullptr);
+  bool is_rsa_private_key = (rsa != nullptr);
+  RSA_free(rsa);
+  BIO_free(in);
+  return is_rsa_private_key;
+}
+
+bool IsEcPrivateKey(absl::string_view private_key_pem) {
+  BIO* in = BIO_new_mem_buf(private_key_pem.data(), private_key_pem.size());
+  if (in == nullptr) {
+    return false;
+  }
+  EC_KEY* ec = PEM_read_bio_ECPrivateKey(in, /*x=*/nullptr,
+                                         /*cb=*/nullptr, /*u=*/nullptr);
```

Comment:
This is a pretty silly because these APIs just call `PEM_read_bio_PrivateKey` and then check if the resulting key is of the expected type. So by calling the two together, you're just repeating work over and over.

Reply:
Thanks for the suggestion! Switched to use `PEM_read_bio_PrivateKey`.